File: 3 - Introduce Polymorphic Creation with Factory Method/C++/before/TestDomBuilder.cpp
#include "TestDomBuilder.h"

#include <gtest/gtest.h>

#include "DomBuilder.h"

void TestDomBuilder::TestAddAboveRoot()
{
  Builder = std::unique_ptr<DomBuilder>(new DomBuilder("orders"));
  ASSERT_NO_THROW(Builder->AddBelow("customer"));
  ASSERT_ANY_THROW(Builder->AddAbove("customer"));
}

File: 3 - Introduce Polymorphic Creation with Factory Method/Python/before/test_dom_builder.py
from unittest import TestCase
from before.dom_builder import DOMBuilder


class TestDOMBuilder(TestCase):
    def test_add_above_root(self):
        invalid_result = """
        <orders>
            <order>
            </order>
        </orders>
        <customer>
        </customer>
        """

        builder = DOMBuilder("orders")
        builder.add_below("order")

        with self.assertRaises(RuntimeError):
            builder.add_above("customer")

